### PR TITLE
[receiver/dockerstats] Add inspect metrics

### PR DESCRIPTION
**Pull request by carlossscastro**
**Description:**

Adds 3 new metrics from the docker inspect api
* container.cpu.limit
* container.cpu.shares
* container.restart.count

Since all these metrics are collected from the inspect api, a new method was created (`recordInspectMetrics`) where only the containerJSON structure is passed as an argument.

**Link to tracking Issue:** #21087

**Testing:**

New metrics were added to the receiver test and expected_metrics.yaml files.
The receiver and config tests were also adjusted to reflect the minimum api version of 1.25

```
❯ go clean -testcache
❯ make test
go test -race -timeout 300s -parallel 4 --tags="" ./...
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver	0.508s
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver/internal/metadata	0.381s
❯ go test --tags integration ./...
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver	14.060s
ok  	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver/internal/metadata	0.241s
```

**Documentation:**
New metrics documentation was generated using `mdatagen`

### CODE REVIEW

**Comment by astencel-sumo** [comment]
Shouldn't `container.restart.count` rather be named `container.restarts`?

According to the [spec](https://github.com/open-telemetry/semantic-conventions/tree/0887921aed44ea6fb818e9dc026b4d8ff8cabe5f/specification/metrics/semantic_conventions#use-count-instead-of-pluralization), `count` should only be used on namespaces. Is `container.restarts` a namespace that can include other metrics? If not, we should use `container.restarts` without the `count` suffix.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (astencel-sumo) [inline receiver/dockerstatsreceiver/metadata.yaml:647]: See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803#issuecomment-1554304700

Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metadata.yaml]: Hi @astencel-sumo 

Thanks for spotting this.
This is not a namespace so it makes all sense to remove the .count. I have submitted the fix (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/3d242fbe92ea61d37cb41090ea81831241a8c240)

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/metadata.yaml:633]: I think all the `container.cpu` metrics should be grouped together here.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/receiver.go:269]: Why `inspect`? is it because these metrics can be gathered via `docker inspect`? If so, I'm not sure if it's a good name for it here, since we aren't using `docker inspect`, we are making API calls directly to the engine.

Perhaps it could be `recordHostConfigMetrics` and you can use the HostConfig type and move the restarts recording out into the calling function.

File: .chloggen/add_inspect_metrics.yaml
Comment (jamesmoessis) [reply .chloggen/add_inspect_metrics.yaml:2]: It's not strictly necessary, but you may want to raise a separate PR for raising the minimum API version, if you want these added metrics to appear purely as an enhancement.

File: receiver/dockerstatsreceiver/receiver.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/receiver.go:270]: Is this approach reliable every time? I'm just asking because I see that we do it differently at Datadog, by relying on multiple attributes. Perhaps on different systems, different properties are available. Take a look please: https://github.com/DataDog/datadog-agent/blob/ab4050c774a8f672fc49c29c4dc5a9a829e45752/pkg/util/containers/metrics/docker/collector_linux.go#L139-L163

**Review by gbbr — COMMENTED** [review]
Thanks for adding this Carlos! Left a small comment, hope it helps :)

File: receiver/dockerstatsreceiver/receiver.go
Comment (gbbr) [reply receiver/dockerstatsreceiver/receiver.go:270]: I think this one is right, check the comment on NanoCPUs: https://github.com/moby/moby/blob/master/api/types/container/hostconfig.go#L330

File: .chloggen/add_inspect_metrics.yaml
Comment (carlossscastro) [reply .chloggen/add_inspect_metrics.yaml:2]: Hi @jamesmoessis,

Yes. Agreed. Did not count on the upgrade for the minimum API version.
I will change this to breaking.

File: receiver/dockerstatsreceiver/receiver.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/receiver.go:129]: @jamesmoessis 
I don't think that is necessary here as Container.JSON is already a pointer
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/internal/docker/docker.go#L30-L33

File: receiver/dockerstatsreceiver/testdata/mock/cgroups_v2/expected_metrics.yaml
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/testdata/mock/cgroups_v2/expected_metrics.yaml:38]: This was just due to using the `golden.WriteMetrics` method to update the expected metrics schemas with the new metrics. It updated also the timestamps for the timestamp when it was regenerated.
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/dockerstatsreceiver/receiver_test.go#L249-L250

File: receiver/dockerstatsreceiver/receiver.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/receiver.go:270]: @gbbr Thanks for pointing that out. I'm currently having a close look at these calculations. Appreciate your comments.

File: receiver/dockerstatsreceiver/receiver.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/receiver.go:269]: Hi @jamesmoessis 
We named it docker inspect following the same naming docker api uses. For those metrics they call it "container inspect" endpoint (https://docs.docker.com/engine/api/v1.42/#tag/Container/operation/ContainerInspect)
But I do agree that the naming is not the best and I do prefer your suggestion of using HostConfig and the restarts, as it is isolated for now, call it directly from the caller function.
Will work on those changes.

File: receiver/dockerstatsreceiver/receiver.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/receiver.go:270]: As @gbbr point out, this is the number of assigned cpus  to the container which can if the format of i.e.`1.5`. Docker stores it in units of 10<sup>9</sup> CPUs.
Dividing by `1e9` will return the exact value set by the user when running the container 
```
docker run -cpus=1.5 ....
```  

For reference: https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metadata.yaml:636]: As per the above reference I would suggest using  `{cpus}`. WDYT?

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metadata.yaml:641]: As per definition on the API:
```
An integer value representing this container's relative CPU weight versus other containers
```

`1024` is the default value. It should be increased or decreased to change the container weight relative to others (https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler)

 So the value here is just an integer therefore we used the unit `1`.

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:129]: Ah, you're right. Makes sense

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:636]: Yes, `{cpus}` makes sense then

File: receiver/dockerstatsreceiver/receiver.go
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver.go:270]: Makes sense, thanks for explaining.

File: .chloggen/add_inspect_metrics.yaml
Comment (gbbr) [inline .chloggen/add_inspect_metrics.yaml:8]: Would it make sense to actually spell out the metrics being added? It may be more useful to users upgrading their Collector. That way, they can immediately start using them rather than starting to dig around to find out their names.

Comment (carlossscastro) [reply .chloggen/add_inspect_metrics.yaml]: Good point @gbbr. Fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/250dd380a647fc29b5722fd2df2b92e7ae20d2da

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (astencel-sumo) [inline receiver/dockerstatsreceiver/metadata.yaml:150]: I don't understand why `container.cpu.shares` is defined as a Sum. If I'm not mistaken, the CPU shares value can be set by the container operator during container creation, just like `container.cpu.limit` - in that case it should also be a Gauge, shouldn't it?

Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metadata.yaml]: Thanks for spotting it @astencel-sumo 

Fix submitted

**Comment by carlossscastro** [comment]
Hi @jamesmoessis

The tests seem to be failing on the k8sattributeprocessor which is unrelated to this PR. Let us know if you need any change from us.

Thanks

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:102]: Isn't it better to not submit this metric if we can't find a limit? Rather than submitting a limit of 0 (zero) ?

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:113]: I wouldn't ignore any errors here. You never know. I would bubble them up and report them.

**Review by gbbr — COMMENTED** [review]
This is looking good! The PR has been open for quite some time now. It would be lovely to see it merged. I just have a few more nits if you'll allow me 🙏🏻

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metric_helper.go:113]: Good point @gbbr. This should really be an edge case as the Docker API checks the syntax of `cpusetCpus` for invalid values, but still it won't hurt bubbling up the error to be safe

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [reply receiver/dockerstatsreceiver/metric_helper.go]: Wow, I was not aware of that possibility. Thanks a lot for taking the time to write the explanation for me. I think you are definitely right and we should keep this as it is. I'll resolve this comment.

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:107]: I would argue that we shouldn't return `nil` here as the error, but return an actual error which says that the CPU limit could not be obtained. As a user, I'd like to know if a metric that I asked for (via configuration) is missing. Otherwise it will be unusual. Maybe at least a warning might be worth it.

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:122]:
```suggestion
			p0, err0 := strconv.Atoi(lineParts[0])
			if err0 != nil {
				return 0, fmt.Errorf("invalid cpusetCpus value: %w", err0)
			}
			p1, err1 := strconv.Atoi(lineParts[1])
			if err1 != nil {
				return 0, fmt.Errorf("invalid cpusetCpus value: %w", err1)
			}
```
I was thinking to suggest wrapping the error, so that it is clear what exactly happened.

**Review by gbbr — COMMENTED** [review]
I hope this helps :)

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metric_helper.go:107]: @gbbr 
Would you think it would be good to return the errors? I don't see this pattern in any other metric and I'm afraid this would create a lot of noise in the logs since it is a common practice not to limit cpu for containers.
What if we added in the metric description something like `This metric is only reported if the container has limits set with -cpus, -cpuset-cpus or -cpu-quota`  
WDYT?

Comment (gbbr) [reply receiver/dockerstatsreceiver/metric_helper.go:107]: I agree. I think that's a good idea.

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:120]:
```suggestion
				return 0, fmt.Errorf("invalid -cpuset-cpus value: %w", err0)
```

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (gbbr) [inline receiver/dockerstatsreceiver/metric_helper.go:124]:
```suggestion
				return 0, fmt.Errorf("invalid -cpuset-cpus value: %w", err1)
```
This way we are referring to the actual flag passed to the CLI, as per your (and Docker official) documentation. Otherwise, the "cpusetCpus" notation doesn't mean much from a user standpoint.

**Review by gbbr — APPROVED** [review]
This LGTM besides the two small nits. @jamesmoessis @astencel-sumo can you please have another look?

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (carlossscastro) [reply receiver/dockerstatsreceiver/metric_helper.go]: Thanks @gbbr

**Comment by carlossscastro** [comment]
@jamesmoessis @astencel-sumo 
Anything else needed to move this forward? Could we get this reviewed?

Thanks

**Comment by jamesmoessis** [comment]
Currently on leave so not able to review anything for a few weeks. @MovieStoreGuy can you please review on my behalf 😄

**Review by MovieStoreGuy — APPROVED** [review]
It looks like all outstanding requests have been made, and I only have one minor comment. 

I will check back on this PR before the next release to make sure there is no concerns, and look at merging.

**Comment by MovieStoreGuy** [comment]
There is a failing test case due to a nil pointer value, could I ask you to take a look at that?

And also rebasing with main as well.

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metric_helper.go]: Changed here: [60bb627](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/60bb62721b24323c58b87ac307c0a34cd273804e)

**Comment by sigilioso** [comment]
I'm taking this over from @carlossscastro. The PR has just been updated from main and the failing test should be working now. Thanks for your review!

**Comment by gbbr** [comment]
@rmfitzpatrick @jamesmoessis I guess we can merge this right?

File: .chloggen/add_inspect_metrics.yaml
Comment (atoulme) [inline .chloggen/add_inspect_metrics.yaml:16]: Can we make this a non-breaking change where we would only mandate version 1.25 if one of the metrics we are adding are enabled?

Comment (sigilioso) [reply .chloggen/add_inspect_metrics.yaml]: Updated here: [58b5270](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/58b5270435394ba7bafeff0b0abd844ebf76ecf3)

**Review by MovieStoreGuy — APPROVED** [review]
These changes look good to me, just a looking for code owner approvals

**Comment by MovieStoreGuy** [comment]
Pinging @jamesmoessis , @rmfitzpatrick

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/metadata.yaml:152]: I'm not sure it's necessary to include `This metric is only reported if the container has limits set with -cpus, -cpuset-cpus or -cpu-quota.` in the reported telemetry considering it must have been set to be reported. Maybe a good candidate for the `if_enabled` warning?

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/metric_helper.go:106]: Does a `-1` default make sense opposed to a zero value?

File: .chloggen/add_inspect_metrics.yaml
Comment (rmfitzpatrick) [inline .chloggen/add_inspect_metrics.yaml:2]: appears the breaking change was lost

File: .chloggen/add_inspect_metrics.yaml
Comment (rmfitzpatrick) [inline .chloggen/add_inspect_metrics.yaml:16]: This isn't conditionally true since the api version check is enforced in config validation.

File: receiver/dockerstatsreceiver/receiver_test.go
Comment (rmfitzpatrick) [inline receiver/dockerstatsreceiver/receiver_test.go:297]: Why would we not want to run the others?

File: receiver/dockerstatsreceiver/metric_helper.go
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metric_helper.go:106]: I'd say `10000` makes sense here, as it is the [default value](https://docs.docker.com/config/containers/resource_constraints/#configure-the-default-cfs-scheduler), we'll get `0` (golang's default) here `hostConfig.CPUPeriod` is not set.

File: .chloggen/add_inspect_metrics.yaml
Comment (sigilioso) [reply .chloggen/add_inspect_metrics.yaml:16]: That's right. We may move it back to breaking or stop enforcing the api version check. WDYT?

File: receiver/dockerstatsreceiver/receiver_test.go
Comment (sigilioso) [reply receiver/dockerstatsreceiver/receiver_test.go]: It's seems a leftover from a local debug session. Thanks for noticing! Updated: [e38b80d](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/e38b80df5ba6cd8e2782020cadca05793e5e65fb)

File: .chloggen/add_inspect_metrics.yaml
Comment (sigilioso) [reply .chloggen/add_inspect_metrics.yaml:2]: It was updated after [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803#discussion_r1332337698). Let's agree on moving it back to breaking or modifying the api version check on the comment bellow if that's ok.

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml:152]: I'm not sure, as the warning message will be:

```
[WARNING] `container.cpu.limit` should not be enabled: This metric is only reported if the container has limits set with -cpus, -cpuset-cpus or -cpu-quota.
```
does it make sense using the extended documentation field instead?

**Review by jamesmoessis — APPROVED** [review]
Thankyou!

File: receiver/dockerstatsreceiver/metadata.yaml
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml]: I took the `extended_documentation approach here: [9b8ec2a](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21803/commits/9b8ec2a97cf6ba01e862cf65064f6069da0745a3), but we can change it if you think a warning is better.

**Comment by ItsLastDay** [comment]
As a bystander, I'm wondering: is it being merged anytime soon? :)

**Comment by jpkrohling** [comment]
@rmfitzpatrick has left a few comments, I'd wait for his approval before merging this.

File: .chloggen/add_inspect_metrics.yaml
Comment (rmfitzpatrick) [reply .chloggen/add_inspect_metrics.yaml:16]: I'm referring to the language in this statement, which isn't accurate.

```suggestion
subtext: It requires API version to 1.25 or greater.
```

**Review by rmfitzpatrick — APPROVED** [review]
lgtm w/ changelog corrections

**Comment by sigilioso** [comment]
While updating from main and fixing conflicts I pushed changes ahead of time and I think that's why all those labels were added, and a lot of folk's review was requested 😓 Sorry for the noise!

**Comment by fatsheep9146** [comment]
here is a failed check
```
Error: ../../receiver/dockerstatsreceiver/receiver.go:124:14: r.recordHostConfigMetrics undefined (type *metricsReceiver has no field or method recordHostConfigMetrics)
Error: ../../receiver/dockerstatsreceiver/receiver.go:287:10: use of package receiver not in selector
```

@sigilioso

**Comment by sigilioso** [comment]
Thanks, I'll take a look ASAP

**Comment by gbbr** [comment]
Why is this PR not getting merged? I see that everything is addressed and it has plenty of approvals.

**Review by MovieStoreGuy — APPROVED** [review]

**Pull request merged**
